# 综合题(选做)

任务要求：白葡萄酒质量共分为7个等级，对白葡萄酒质量进行预测。
1. 使用精度作为评价指标。
2. 使用一元线性回归、多元线性回归、对数线性回归等线性回归模型对葡萄酒质量进行预测，计算其十折交叉验证的精度。
3. 使用对数几率回归、线性判别分析对葡萄酒质量进行预测，计算其十折交叉验证的精度。
4. 对比不同模型的预测能力。

## 读取数据

In [3]:
import pandas as pd
import numpy as np
data = pd.read_csv('data/wine_quality/winequality-white.csv', delimiter=";")

In [4]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [6]:
# YOUR CODE HERE
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_predict
from sklearn.linear_model import LinearRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
import math

datax = data.iloc[:,:11]
datay = data.iloc[:,11:]

data1x = datax.iloc[:,0:1]

model1 = LinearRegression()
model1_prediction = cross_val_predict(model1,data1x,datay,cv=10)
print("一元线性回归：")
model1_prediction = np.rint(model1_prediction)
print(accuracy_score(datay.values.flatten(),pd.DataFrame(model1_prediction).values.flatten()))

model2_prediction = cross_val_predict(model1,datax,datay,cv=10)
print("多元线性回归：")
model2_prediction = np.rint(model2_prediction)
print(accuracy_score(datay.values.flatten(),pd.DataFrame(model2_prediction).values.flatten()))

data1y = pd.DataFrame(np.log(np.array(datay)))
model3_prediction = cross_val_predict(model1,datax,data1y,cv=10)
print("对数线性回归：")
model3_prediction = np.rint(np.exp(model3_prediction))
print(accuracy_score(datay.values.flatten(),pd.DataFrame(model3_prediction).values.flatten()))

model4 = LogisticRegression()
model4_prediction = cross_val_predict(model4,datax,datay,cv=10)
print("对数几率回归：")
model4_prediction = np.rint(model4_prediction)
print(accuracy_score(datay.values.flatten(),pd.DataFrame(model4_prediction).values.flatten()))

model5 = LinearDiscriminantAnalysis()
model5_prediction = cross_val_predict(model5,datax,datay,cv=10)
print("线性判别分析：")
model5_prediction = np.rint(model5_prediction)
print(accuracy_score(datay.values.flatten(),pd.DataFrame(model5_prediction).values.flatten()))

一元线性回归：
0.4481420988158432
多元线性回归：
0.5132707227439771
对数线性回归：
0.5208248264597795


c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 5 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataCon

对数几率回归：
0.5228664761126991
线性判别分析：
0.5212331563903634


c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 5 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:578: DataCon

## 预测能力对比
模型|精度
-|-
一元线性回归 | 0.4481420988158432
多元线性回归 | 0.5132707227439771
对数线性回归 | 0.5208248264597795
对数几率回归 | 0.5228664761126991
线性判别回归 | 0.5212331563903634
